Optionally return a StreamConnectionPair in StreamClientConnection::create

donny-dont opened the pull request:
#### fd3a4d231104825ad39d97741e09928f391c32bb
<pre>
Optionally return a StreamConnectionPair in StreamClientConnection::create
<a href="https://bugs.webkit.org/show_bug.cgi?id=261782">https://bugs.webkit.org/show_bug.cgi?id=261782</a>

Reviewed by Kimmo Kinnunen.

Change function signature of `StreamClientConnection::create` to return
a `std::optional&lt;StreamConnectionPair&gt;`. Modify `StreamConnectionPair`
to use a `Ref&lt;StreamClientConnection&gt;` since the value will never be
`null`. Updated the call sites for `StreamClientConnection::create` to
handle the changes.

This is a step towards `SharedMemory::Handle` also always being valid.

* Source/WebKit/Platform/IPC/StreamClientConnection.cpp:
(IPC::StreamClientConnection::create):
* Source/WebKit/Platform/IPC/StreamClientConnection.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.cpp:
(WebKit::RemoteGraphicsContextGLProxy::create):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::ensureGPUProcessConnection):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteGPUProxy.cpp:
(WebKit::RemoteGPUProxy::create):
* Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
(WebKit::IPCTestingAPI::createError):
(WebKit::IPCTestingAPI::JSIPC::createStreamClientConnection):
* Tools/TestWebKitAPI/Tests/IPC/StreamConnectionTests.cpp:

Canonical link: <a href="https://commits.webkit.org/268220@main">https://commits.webkit.org/268220@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0ab105e6c7f37568e03afcb72252eb2f8ce8a82a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/19023 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/19367 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/19975 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/20889 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17786 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/19210 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/22660 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/19493 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/19545 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/19247 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/19346 "4 failures") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/16553 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/21776 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/16555 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/17336 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/23734 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/17594 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/17508 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21680 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/18085 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/15329 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/17172 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4535 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/21531 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/17920 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.cpp:
```diff
@@ -81,15 +81,15 @@ RefPtr<RemoteGraphicsContextGLProxy> RemoteGraphicsContextGLProxy::create(IPC::C
-    auto [clientConnection, serverConnectionHandle] = IPC::StreamClientConnection::create(connectionBufferSizeLog2);
-    if (!clientConnection)
+    auto connectionPair = IPC::StreamClientConnection::create(connectionBufferSizeLog2);
+    if (!connectionPair || !connectionPair->streamConnection)
```
so here the contract must be that either connetionPair is nullopt or then it's fully usable.
So checking for !connectionPair should be enough.

kkinnunen-apple commented on Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.cpp:
```diff
@@ -81,15 +81,15 @@ RefPtr<RemoteGraphicsContextGLProxy> RemoteGraphicsContextGLProxy::create(IPC::C
+    if (!connectionPair || !connectionPair->streamConnection)
         return nullptr;
-    auto instance = platformCreate(connection, clientConnection.releaseNonNull(), attributes
+    auto instance = platformCreate(connection, connectionPair->streamConnection.releaseNonNull(), attributes
```
in the if continuation you could just deconstruct like the previous code:
```
auto [clientConnectin, serverConnectionHandle] = *connectionPair;
```

This would avoid the long names.

donny-dont replied:
Just a heads up that I didn't see that compiling but the following it was happy with.

```cpp
auto [clientConnectin, serverConnectionHandle] = WTFMove(*connectionPair);
```

kkinnunen-apple commented on Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
```diff
@@ -85,15 +85,15 @@ void RemoteRenderingBackendProxy::ensureGPUProcessConnection()
         static constexpr auto connectionBufferSizeLog2 = 21;
-        auto [streamConnection, serverHandle] = IPC::StreamClientConnection::create(connectionBufferSizeLog2);
-        if (!streamConnection)
+        auto connectionPair = IPC::StreamClientConnection::create(connectionBufferSizeLog2);
```
same if comment and the deconstruction comment

kkinnunen-apple commented on Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteGPUProxy.cpp:
```diff
@@ -46,11 +46,11 @@ namespace WebKit {
     constexpr size_t connectionBufferSizeLog2 = 21;
-    auto [clientConnection, serverConnectionHandle] = IPC::StreamClientConnection::create(connectionBufferSizeLog2);
-    if (!clientConnection)
+    auto connectionPair = IPC::StreamClientConnection::create(connectionBufferSizeLog2);
```
same if comment and deconstruction comment.

kkinnunen-apple commented on Source/WebKit/Platform/IPC/StreamClientConnection.h:
```diff
@@ -63,7 +63,7 @@ class StreamClientConnection final : public ThreadSafeRefCounted<StreamClientCon
     };
 
     // The messages from the server are delivered to the caller through the passed IPC::MessageReceiver.
```
Above, you could make it `Ref<StreamClientConnection> streamConnection` since it cannot be non-null.

kkinnunen-apple commented on Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
```diff
@@ -2475,15 +2475,19 @@ JSValueRef JSIPC::createStreamClientConnection(JSContextRef context, JSObjectRef
     }
     auto connectionPair = IPC::StreamClientConnection::create(*bufferSizeLog2);
+    if (!connectionPair) {
+        *exception = createTypeError(context, "?????????"_s);
```
you could prototype with following:

This near the createTypeError:
```
static JSValueRef createError(JSContextRef context, const String& message)
{
    JSC::JSLockHolder lock(toJS(context)->vm());
    return toRef(JSC::createError(toJS(context), message));
}
```

then use something like `createError(context, "Failed to create the connection"_s)`

kkinnunen-apple reviewed (approved):
Some suggestions

kkinnunen-apple commented on Tools/TestWebKitAPI/Tests/IPC/StreamConnectionTests.cpp:
```diff
@@ -212,32 +212,32 @@ class StreamConnectionTest : public ::testing::Test, public StreamConnectionTest
-    auto [clientConnection, serverConnectionHandle] = IPC::StreamClientConnection::create(defaultBufferSizeLog2);
-    ASSERT_TRUE(clientConnection);
-    auto serverConnection = IPC::StreamServerConnection::tryCreate(WTFMove(serverConnectionHandle)).releaseNonNull();
+    auto connectionPair = = IPC::StreamClientConnection::create(defaultBufferSizeLog2);
```
this = = is your compile error.

donny-dont replied:
Copy pasta error....

donny-dont committed:
Optionally return a StreamConnectionPair in StreamClientConnection::create
https://bugs.webkit.org/show_bug.cgi?id=261782

Reviewed by Kimmo Kinnunen.

Change function signature of `StreamClientConnection::create` to return
a `std::optional<StreamConnectionPair>`. Modify `StreamConnectionPair`
to use a `Ref<StreamClientConnection>` since the value will never be
`null`. Updated the call sites for `StreamClientConnection::create` to
handle the changes.

This is a step towards `SharedMemory::Handle` also always being valid.

* Source/WebKit/Platform/IPC/StreamClientConnection.cpp:
(IPC::StreamClientConnection::create):
* Source/WebKit/Platform/IPC/StreamClientConnection.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.cpp:
(WebKit::RemoteGraphicsContextGLProxy::create):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::ensureGPUProcessConnection):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteGPUProxy.cpp:
(WebKit::RemoteGPUProxy::create):
* Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
(WebKit::IPCTestingAPI::createError):
(WebKit::IPCTestingAPI::JSIPC::createStreamClientConnection):
* Tools/TestWebKitAPI/Tests/IPC/StreamConnectionTests.cpp:

Canonical link: https://commits.webkit.org/268220@main
